In [1]:
from pulp import *
import pandas as pd

In [51]:
def bakery(sugar,chocolate):
    prob = LpProblem('Bakery', LpMaximize)#model
    x1 = LpVariable('x1', lowBound=0)#non negativity constraints
    x2 = LpVariable('x2', lowBound=0)#non negativity constraints
    x3 = LpVariable('x3', lowBound=0)#non negativity constraints

    prob += 3*x1 + 7*x2 + 5*x3 #objective function
    prob += x1 + x2 + x3 <= sugar #sugar constraint
    prob += 2*x1 + 3*x2 + x3 <= chocolate #chocolate constraint

    print(prob)
    prob.solve()# linear programming solution

    print(f"reduced cost x1:{value(x1)}")
    print(f"reduced cost x2:{value(x2)}")
    print(f"reduced cost x3:{value(x3)}")
    print(f"optimized objective function: {value(prob.objective)}")
    
    for name, c in prob.constraints.items():
       print( name, ":", c, "\t", f' shadow price {c.pi}', "\t\t", f' slack variables {c.slack}')
  
bakery(sugar = 50,chocolate = 100)   

Bakery:
MAXIMIZE
3*x1 + 7*x2 + 5*x3 + 0
SUBJECT TO
_C1: x1 + x2 + x3 <= 50

_C2: 2 x1 + 3 x2 + x3 <= 100

VARIABLES
x1 Continuous
x2 Continuous
x3 Continuous

reduced cost x1:0.0
reduced cost x2:25.0
reduced cost x3:25.0
optimized objective function: 300.0
_C1 : x1 + x2 + x3 <= 50 	  shadow price 4.0 		  slack variables -0.0
_C2 : 2*x1 + 3*x2 + x3 <= 100 	  shadow price 1.0 		  slack variables -0.0


To maximize profit and obtain the optimal solution, Of item 1 , zero should be produced, while of item 2 and 3  25 should be produced, hence more resources should be focused on item 2 and 3 and the focus shouldnt be on non negative and zero co-efficients

In [52]:
bakery(sugar = 51,chocolate = 100)   

Bakery:
MAXIMIZE
3*x1 + 7*x2 + 5*x3 + 0
SUBJECT TO
_C1: x1 + x2 + x3 <= 51

_C2: 2 x1 + 3 x2 + x3 <= 100

VARIABLES
x1 Continuous
x2 Continuous
x3 Continuous

reduced cost x1:0.0
reduced cost x2:24.5
reduced cost x3:26.5
optimized objective function: 304.0
_C1 : x1 + x2 + x3 <= 51 	  shadow price 4.0 		  slack variables -0.0
_C2 : 2*x1 + 3*x2 + x3 <= 100 	  shadow price 1.0 		  slack variables -0.0


In [53]:
bakery(sugar = 50,chocolate = 101)   

Bakery:
MAXIMIZE
3*x1 + 7*x2 + 5*x3 + 0
SUBJECT TO
_C1: x1 + x2 + x3 <= 50

_C2: 2 x1 + 3 x2 + x3 <= 101

VARIABLES
x1 Continuous
x2 Continuous
x3 Continuous

reduced cost x1:0.0
reduced cost x2:25.5
reduced cost x3:24.5
optimized objective function: 301.0
_C1 : x1 + x2 + x3 <= 50 	  shadow price 4.0 		  slack variables -0.0
_C2 : 2*x1 + 3*x2 + x3 <= 101 	  shadow price 1.0 		  slack variables -0.0


An additional unit of sugar yielded a higher maximized profit at 304 vs an additional unit of chocolate at 301, so sugar will be the preferred choice and the values corresponds to the shadow price output 

In [54]:
def advertising(women, men):


    prob2 = LpProblem('Advertising', LpMinimize) #model
    m1 = LpVariable('m1', lowBound=0)#non negativity constraints
    m2 = LpVariable('m2', lowBound=0)#non negativity constraints


    prob2 += 50*m1 + 100*m2 #objective function
    prob2 += 7*m1 + 2*m2  >= women #women constraint
    prob2 += 2*m1 + 12*m2 >= men #men constraint

    print(prob2)
    prob2.solve()

    print(f"reduced cost x1:{value(m1)}")
    print(f"reduced cost x2:{value(m2)}")
    print(f"optimized objective function: {value(prob2.objective)}")
    for name, c in prob2.constraints.items():
       print( name, ":", c, "\t", f' shadow price {c.pi}', "\t\t", f' slack variables {c.slack}')
advertising(women = 28, men = 24)

Advertising:
MINIMIZE
50*m1 + 100*m2 + 0
SUBJECT TO
_C1: 7 m1 + 2 m2 >= 28

_C2: 2 m1 + 12 m2 >= 24

VARIABLES
m1 Continuous
m2 Continuous

reduced cost x1:3.6
reduced cost x2:1.4
optimized objective function: 320.0
_C1 : 7*m1 + 2*m2 >= 28 	  shadow price 5.0 		  slack variables -0.0
_C2 : 2*m1 + 12*m2 >= 24 	  shadow price 7.5 		  slack variables -0.0


In [49]:
advertising(women = 27, men = 24)

Advertising:
MINIMIZE
50*m1 + 100*m2 + 0
SUBJECT TO
_C1: 7 m1 + 2 m2 >= 27

_C2: 2 m1 + 12 m2 >= 24

VARIABLES
m1 Continuous
m2 Continuous

reduced cost x1:3.45
reduced cost x2:1.425
optimized objective function: 315.0
_C1 : 7*m1 + 2*m2 >= 27 	  shadow price 5.0 		  slack variables -0.0
_C2 : 2*m1 + 12*m2 >= 24 	  shadow price 7.5 		  slack variables -0.0


In [50]:
advertising(women = 28, men = 23)

Advertising:
MINIMIZE
50*m1 + 100*m2 + 0
SUBJECT TO
_C1: 7 m1 + 2 m2 >= 28

_C2: 2 m1 + 12 m2 >= 23

VARIABLES
m1 Continuous
m2 Continuous

reduced cost x1:3.625
reduced cost x2:1.3125
optimized objective function: 312.5
_C1 : 7*m1 + 2*m2 >= 28 	  shadow price 5.0 		  slack variables -0.0
_C2 : 2*m1 + 12*m2 >= 23 	  shadow price 7.5 		  slack variables -0.0


1 million less men yielded a minimized cost at 312.5 vs an 1 million less women at 315 which corresponds to the shadow price output and to save money 1 million less men will yield a minimized cost